### PR TITLE
Don't warn if cloud-init is disabled

### PR DESCRIPTION
**Pull request by holmanb**
## Proposed Commit Message
Use individual commit messages

## Additional Context
Also relocate apport functionality to main branch for better testing and less code duplication.

Fixes https://github.com/canonical/cloud-init/issues/4875, which is [this bug](https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/2025376) which was never fixed in `ubuntu/focal` so it came back again in 23.3.

If we had included the apport code in the main branch rather than in packaging branches, our unit tests would have caught the invalid typing.

## Merge type

- [ ] Squash merge using "Proposed Commit Message"
- [x] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: cloudinit/apport.py [hunk 1]
```diff
@@ -226,3 +240,102 @@ def add_info(report, ui):
...
+    """Entry point for Apport.
+
+    This hook runs for every apport report
+
+    Add a subset of non-sensitive cloud-init data from
+    /run/cloud/instance-data.json that will be helpful for debugging.
+    """
+    try:
+        if not is_cloud_init_enabled():
+            return
```
Comment (blackboxsw) [inline cloudinit/apport.py:309]: Confirmed that this is the only diff of substance between the downstream packaged debian/apport-general-hooks.py and our upstream version, I approve the "new/migrated" function additions.

File: cloudinit/apport.py [hunk 2]
```diff
@@ -71,6 +72,13 @@
 CLOUDINIT_OUTPUT_LOG = "/var/log/cloud-init-output.log"
 
 
+def is_cloud_init_enabled() -> bool:
+    return (
+        get_status_details(read_cfg_paths()).boot_status_code
+        not in DISABLED_BOOT_CODES
+    )
```
Comment (blackboxsw) [inline cloudinit/apport.py:79]: Given that this is more status-related functionality than apport-related It is worth just defining this helper function in cloudinit.status?

**Review by blackboxsw — APPROVED** [review]
Minor request to mograte is_cloud_init_enabled into cloudinit.status module so we aren't importing multiple "implementation details" over into apport. But +1 either way if you disagree with that suggestion.

File: cloudinit/apport.py (hunk 2 above)
Comment (holmanb) [reply cloudinit/apport.py]: That works for me!

Comment (holmanb) [reply cloudinit/apport.py]: Aside: It does feel funny to have code under `cloudinit/cmd/` that ends up getting imported by external tools.

I would almost lean towards a `cloudinit/lib/` to have documented acceptable interfaces for external users with a very small subset of functions that we guarantee to have a stable interface. But that's very much a "someday" idea - for now I think that your suggested approach is best. Thanks for the review @blackboxsw!

**Pull request merged**